### PR TITLE
Add global batch-wise MoE auxiliary loss

### PR DESCRIPTION
**Pull request by pianpwk**
microbatching supported with algo from https://arxiv.org/abs/2501.11873

Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #4772

### CODE REVIEW

File: torchtitan/models/gpt_oss/__init__.py [hunk 1]
```diff
@@ -253,7 +256,7 @@ def _debugmodel(
             hidden_dim=hidden_dim,
             num_experts=8,
             top_k=4,
-            load_balance_coeff=1e-3,
```
Comment (tianyu-l) [inline torchtitan/models/gpt_oss/__init__.py:256]: did you find out which load balancing scheme gpt-oss use from literature?

File: torchtitan/models/qwen3/__init__.py [hunk 2]
```diff
@@ -187,6 +188,9 @@ def _build_qwen3_moe_layers(
                         comm_backend=moe_comm_backend,
                         non_blocking_capacity_factor=non_blocking_capacity_factor,
                     ),
+                    aux_loss_coeff=1e-3,
```
Comment (tianyu-l) [inline torchtitan/models/qwen3/__init__.py:191]: I think we also need to make this None for qwen
https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/common/config_utils.py#L316

Actually default should be `None` to avoid mistake in the future.

File: torchtitan/models/qwen3/__init__.py [hunk 3]
```diff
@@ -586,6 +590,12 @@ def _235b_a22b(
 }
 
 
+def _post_optimizer_build_fn(optimizers, model_parts, parallel_dims):
+    """Register step pre-hooks for load balancing and aux-loss accumulators."""
+    register_moe_load_balancing_hook(optimizers, model_parts, parallel_dims)
```
Comment (tianyu-l) [inline torchtitan/models/qwen3/__init__.py:595]: don't need this?

File: torchtitan/models/common/config_utils.py [hunk 4]
```diff
@@ -315,13 +316,21 @@ def make_moe_config(
...
+    aux_loss = None
     if aux_loss_coeff is not None:
-        router = dataclasses.replace(
-            router,
-            aux_loss=MicrobatchWiseLoadBalanceLoss.Config(coeff=aux_loss_coeff),
-        )
+        if aux_loss_type == "microbatch_wise":
+            aux_loss = MicrobatchWiseLoadBalanceLoss.Config(coeff=aux_loss_coeff)
+        elif aux_loss_type == "global_batch_wise":
+            aux_loss = GlobalBatchWiseLoadBalanceLoss.Config(coeff=aux_loss_coeff)
```
Comment (tianyu-l) [inline torchtitan/models/common/config_utils.py:327]: sounds unambiguous to just
```suggestion
        elif aux_loss_type == "batch_wise":
            aux_loss = BatchWiseLoadBalanceLoss.Config(coeff=aux_loss_coeff)
```

File: torchtitan/models/common/moe.py [hunk 5]
```diff
@@ -511,6 +511,37 @@ def _init_self_buffers(self, *, buffer_device: torch.device | None = None) -> No
             )
 
 
+def _reduce_token_partials(
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:514]: does it make sense to put them under `AuxLoss` for now?

File: torchtitan/models/common/moe.py [hunk 6]
```diff
@@ -613,33 +619,101 @@ def forward(
...
+        # rejected by spmd_types.
+        counts_E = routing_map_TE.to(scores_TE.dtype).sum(dim=0)
+
+        # Counts define the non-differentiable global routing frequency, so
+        # every DP coordinate must use the same value.
+        global_counts_E = _reduce_token_partials(counts_E, include_dp_axis=True)
+        if self._cumulative_expert_counts_E is None:
+            self._cumulative_expert_counts_E = torch.zeros_like(
+                global_counts_E, dtype=torch.float32
+            )
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:692]: may not be cudagraph friendly if we reuse cudagraph for every spmd (outer) grad accum? But I think we probably will move to all spmd grad accum steps captured in one big cudagraph anyway. cc @fegin 

That said, any reason we don't allocate memory for the final shape upfront?

File: torchtitan/config/validation.py [hunk 7]
```diff
@@ -14,7 +14,31 @@
...
+        return
+
+    if num_microbatches_per_step > 1 and activation_checkpoint_enabled:
+        # TODO: Once activation checkpointing has migrated to torch_remat, use
+        # remat.is_recomputing() to keep replay from advancing rolling counts.
+        raise ValueError(
+            "GlobalBatchWiseLoadBalanceLoss with multiple microbatches is "
+            "incompatible with activation checkpointing because recomputation "
+            "would advance its rolling expert counts. Disable activation "
+            "checkpointing or use MicrobatchWiseLoadBalanceLoss."
```
Comment (tianyu-l) [inline torchtitan/config/validation.py:40]: Sounds too big of a limitation, as models using this loss cannot be trained with AC anymore.

On the other hand, I believe torch_remat can work around this? @anijain2305

**Review by tianyu-l — CHANGES_REQUESTED** [review]
